sources/core/Stride.Core/Diagnostics/ProfilingState.cs

```diff
@@ -142,27 +87,22 @@ public void Begin(string textFormat, params object[] textFormatArguments)
         /// <param name="value1">Second value (can be int, float, long or double).</param>
         /// <param name="value2">Third value (can be int, float, long or double).</param>
         /// <param name="value3">Fourth value (can be int, float, long or double).</param>
-        public void Begin(string text, ProfilingCustomValue? value0, ProfilingCustomValue? value1 = null, ProfilingCustomValue? value2 = null, ProfilingCustomValue? value3 = null)
+        public void Begin(string text, ProfilingCustomValue? value0 = null, ProfilingCustomValue? value1 = null, ProfilingCustomValue? value2 = null, ProfilingCustomValue? value3 = null)
         {
-            EmitEvent(ProfilingMessageType.Begin, text, value0, value1, value2, value3);
+            EmitEvent(ProfilingMessageType.Begin, new ProfilingEventMessage(text, value0, value1, value2, value3));
         }
 
         /// <summary>
-        /// Emits a Begin event with the specified formatted text.
+        /// Emits a Begin event with an override on the timestamp. Internal for the use of Stride.Rendering.
         /// </summary>
-        /// <param name="value0">First value (can be int, float, long or double).</param>
-        /// <param name="value1">Second value (can be int, float, long or double).</param>
-        /// <param name="value2">Third value (can be int, float, long or double).</param>
-        /// <param name="value3">Fourth value (can be int, float, long or double).</param>
-        public void Begin(ProfilingCustomValue? value0, ProfilingCustomValue? value1 = null, ProfilingCustomValue? value2 = null, ProfilingCustomValue? value3 = null)
+        internal void BeginGpu(long timeStamp, long tickFrequency)
         {
-            EmitEvent(ProfilingMessageType.Begin, null, value0, value1, value2, value3);
-        }
+            // Perform event only if the profiling is running (to save on time calculations)
+            if (!isEnabled) return;
 
-        internal void Begin(long timeStamp)
-        {
+            this.tickFrequency = tickFrequency > 0 ? tickFrequency : Stopwatch.Frequency;
```

Comment:
Shouldn't this throw instead ? If the frequency tends to be invalid, I would think it's the caller's responsibility to ensure that the frequency is valid ?

Reply:
It's confusing because I didn't use `this.` on the right side - basically if the `tickFrequency` has been set we keep it and if not we fall back to the Stopwatch frequency. I might use a nullable field instead ??= assignment here.

Reply:
> basically if the tickFrequency has been set we keep it

Okay, maybe I'm confused indeed, but afaik, this is what this code does:
``thisFreq = paramFreq > 0 ? paramFreq : Stopwatch.Frequency``
From what you've just said, this is what I think you meant to do:
``thisFreq = thisFreq > 0 ? paramFreq : Stopwatch.Frequency``
Did I understand correctly ?

Reply:
Hmm, no, my mistake (was responding based on diff and missed the param to that method).
Now even I got a little confused. I'll try to make this a bit more clear.
The basic issue is that TickFrequency is Stopwatch based everywhere except for GPU events where it's set on BeginGpu but needs to be available for EndGpu.
I decided to move this set into a property explicitly.